src/day1.rs

```diff
@@ -0,0 +1,57 @@
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::fs;
+
+// Challenge text: https://adventofcode.com/2022/day/1
+//
+// You're given a file with a groups of numbers, one per line, with different
+// groups separated by an empty line. Part 1 asks for the sum of numbers from
+// the group with highest sum; part 2 asks for the sum of the 3 groups with the
+// highest sums.
+
+// elf_iterator takes a string comprised of groups of numbers (one per line),
+// separated by empty lines, and returns an iterator over the sum of the numbers
+// in each group.
+//
+// For example, for the input:
+//  1
+//  2
+//
+//  3
+//  4
+//  5
+//
+// it will returns an iterator over {3, 12}.
+fn elf_iterator<'a>(contents: &'a str) -> impl Iterator<Item = i32>  +  'a  {
```

Comment:
I had to take `contents` as an arg, and templatize on its lifetime, because `groups.map()` borrows from `contents`. This make sense. But if I wanted `contents` to be defined inside this function, what options would I have had to somehow embed it into the return value so it can be dropped by the caller?

Reply:
The key point is that the type `&str` is a reference, so you can't get something that owns it. There are two ways that you could define `contents` inside this function:
- use a `String` which is an owned type and return an iterator that owns it by value. This conversation is provided by the `IntoIterator` trait.
- use a constant `&str`. This would be a reference with a static lifetime, so returning a borrowed type would work; you'll need to tell the compiler that by annotating the return with `'static`

Come to think of it, since you're returning `impl ...` you'll need the 'static lifetime on the return value in both cases.

Reply:
> use a String which is an owned type and return an iterator that owns it by value. This conversation is provided by the IntoIterator trait.

I can't figure out this suggestion exactly. Can you spell it out?
I can do something like this:
```
    let contents: String = fs::read_to_string("input/day1.txt").expect("failed to read input file");
    let groups: Vec<String> = contents.split("\n\n").map(String::from).collect();
    groups.into_iter().map(|g| { 
	...
```
but this is not satisfying because I'm consuming all of the `contents.split()` iterator, storing all the items. I'd like to avoid that; I'd like to not store more than one element at a time (apart from the whole `contents` String). So, I want to return an iterator over parts of `contents`, and also return all of `contents`, all bundled together with the same lifetime. I assume I can do that if I define my own type that owns `contents` and also exposes the iterator over it, but can I do it without defining my own type?
Does this make sense?

Reply:
You probably cannot do that without making your own type because of the signature of [`String::split`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.split) which takes `&self` (so it can't possible own the receiver).

If you define your own type for this purpose, it will need to `impl Iterator`; in other words, this can't be done by composing the iterator returned from [`String::split`] because the containing structure would become self-referential, which means it can't be moved, which is a problem in Rust where everything is move by default. You can use `std::pin::Pin` to tell the type system that your type is self-referential and may not be moved, but then you wouldn't be able to return it from this function.